File: 2015/2/c++/main.cpp
#include <iostream>
#include <string>
#include <fstream>

int main(int argc, char **argv)
{
    if (argc < 2)
    {
        std::cout << "Usage: " << argv[0] << " <input_file>" << std::endl;
        return 1;
    }

    std::string file_name = argv[1];

    std::ifstream file(file_name);

    if (!file.is_open())
    {
        std::cout << "Error: Could not open file " << file_name << std::endl;
        return 1;
    }

    int total = 0;
    std::string line;
    while (std::getline(file, line))
    {
        int a, b, c;
        sscanf(line.c_str(), "%dx%dx%d", &a, &b, &c);

        int area = 2 * a * b + 2 * b * c + 2 * c * a;

        int smallest = a * b;
        if (b * c < smallest)
        {
            smallest = b * c;
        }
        if (c * a < smallest)
        {
            smallest = c * a;
        }

        area += smallest;
        total += area;
    }

    std::cout << total << std::endl;

    file.close();

    return EXIT_SUCCESS;
}

File: 2015/ruby/2/main.rb
#!/usr/bin/env ruby

def main()
    args = ARGV

    if args.length < 1
        puts "Usage: ruby main.rb <input.txt>"
        exit 1
    end

    filename = args[0]
    file = File.open(filename, "r")

    if file.nil?
        puts "Could not open file"
        exit 1
    end

    lines = file.readlines
    file.close

    paper_size = 0
    ribbon_length = 0

    lines.each do |line|
        dimensions = line.split('x').map(&:to_i)
        dimensions.sort!
        l, w, h = dimensions

        paper_size += 2*l*w + 2*w*h + 2*h*l + l*w
        ribbon_length += 2*l + 2*w + l*w*h
    end

    puts "Part 1: #{paper_size}" 
    puts "Part 2: #{ribbon_length}"
end 

main()